ext/fs/ops.rs

```diff
@@ -115,23 +115,48 @@ fn print_not_capable_info(standalone: bool, err: &'static str) -> String {
   }
 }
 
+// workaround for rustc incorrectly inferring closure type
+fn cast<T>(t: T) -> T
+where
+  T: for<'a> FnMut(
+    bool,
+    &'a Path,
+    &'a OpenOptions,
+  ) -> deno_io::fs::FsResult<std::borrow::Cow<'a, Path>>,
+{
+  t
+}
+
 fn sync_permission_check<'a, P: FsPermissions + 'static>(
   permissions: &'a mut P,
   api_name: &'static str,
-) -> impl AccessCheckFn + 'a {
-  move |resolved, path, options| {
-    permissions.check(resolved, options, path, api_name)
+) -> Option<impl AccessCheckFn + 'a> {
+  if permissions.allows_all() {
+    None
+  } else {
+    Some(cast(move |resolved, path, options| {
+      permissions.check(resolved, options, path, api_name)
+    }))
   }
 }
 
 fn async_permission_check<P: FsPermissions + 'static>(
   state: Rc<RefCell<OpState>>,
   api_name: &'static str,
-) -> impl AccessCheckFn {
-  move |resolved, path, options| {
-    let mut state = state.borrow_mut();
-    let permissions = state.borrow_mut::<P>();
-    permissions.check(resolved, options, path, api_name)
+) -> Option<impl AccessCheckFn> {
+  let allows_all = {
+    let mut permissions = state.borrow_mut();
+    let permissions = permissions.borrow_mut::<P>();
+    permissions.allows_all()
+  };
+  if allows_all {
+    None
```

Comment:
How hard is it to move the canonicalization stuff into `FsPermissions::check_open`? Ideally we'd only be accessing `Permissions` once from the OpState and then aquiring its lock once instead of now twice when not running with `-A`.

Reply:
moved things around. essentially changed it to pass an argument to the permission check that let's you get the resolved path